[Bug] Cannot serialize unique_ptr with virtual inheritance of a PUP::able

This is a regression from https://github.com/UIUC-PPL/charm/pull/3599 I don't understand why the code change there was made, but it caused a regression for us. We now cannot build if we have:

```
    PUP::able
  /   \
/       \
A      B
 \     /
  \   /
   C
````
we get the following error:
```
In file included from /home/nils/SpECTRE/spectre/tests/Unit/PointwiseFunctions/AnalyticData/Xcts/Test_Binary.cpp:4:
In file included from /home/nils/SpECTRE/spectre/tests/Unit/Framework/TestingFramework.hpp:27:
In file included from /home/nils/SpECTRE/spectre/src/Utilities/System/Exit.hpp:6:
In file included from /home/nils/SpECTRE/charm/multicore-linux-x86_64-gcc/include/charm++.h:24:
In file included from /home/nils/SpECTRE/charm/multicore-linux-x86_64-gcc/include/cklists.h:6:
/home/nils/SpECTRE/charm/multicore-linux-x86_64-gcc/include/pup_stl.h:477:15: error: cannot cast 'PUP::able *' to 'elliptic::analytic_data::AnalyticSolution *' via virtual base 'PUP::able'
      t.reset((T*)t1);
              ^~~~~~
/home/nils/SpECTRE/charm/multicore-linux-x86_64-gcc/include/pup_stl.h:486:5: note: in instantiation of function template specialization 'PUP::pup<elliptic::analytic_data::AnalyticSolution, nullptr>' requested here
    pup(p, t);
    ^
/home/nils/SpECTRE/charm/multicore-linux-x86_64-gcc/include/pup_stl.h:415:54: note: in instantiation of function template specialization 'PUP::operator|<elliptic::analytic_data::AnalyticSolution>' requested here
    std::for_each(a.begin(), a.end(), [&p](T& t) { p | t; });
                                                     ^
/home/nils/SpECTRE/charm/multicore-linux-x86_64-gcc/include/pup_stl.h:426:5: note: in instantiation of function template specialization 'PUP::pup<std::unique_ptr<elliptic::analytic_data::AnalyticSolution>, 2UL, nullptr>' requested here
    pup(p, a);
    ^
/home/nils/SpECTRE/spectre/src/PointwiseFunctions/AnalyticData/Xcts/Binary.hpp:364:7: note: in instantiation of function template specialization 'PUP::operator|<std::unique_ptr<elliptic::analytic_data::AnalyticSolution>, 2UL>' requested here
    p | superposed_objects_;
      ^
/home/nils/SpECTRE/spectre/tests/Unit/PointwiseFunctions/AnalyticData/Xcts/Test_Binary.cpp:85:23: note: in instantiation of member function 'Xcts::AnalyticData::Binary<elliptic::analytic_data::AnalyticSolution, brigand::list<Xcts::Solutions::Schwarzschild>>::pup' requested here
  const auto binary = serialize_and_deserialize(derived);
                      ^
1 error generated.
```
I believe `pup` needs to be called "bottom up" for it to be safe (the opposite order of construction) which is also how I interpret the Charm++ manual https://charm.readthedocs.io/en/latest/charm%2B%2B/manual.html#subclass-allocation-via-pup-able with the example that every _derived_ class needs to first call its base class `pup` (construction order).

--- a/src/util/pup_stl.h
+++ b/src/util/pup_stl.h
@@ -474,9 +474,9 @@ using Requires = typename requires_impl<
     PUP::able* t1 = nullptr;
     if (p.isUnpacking()) {
       p | t1;
-      t.reset((T*)t1);
+      t.reset(dynamic_cast<T*>(t1));
     } else {
-      t1 = (PUP::able*)t.get();
+      t1 = dynamic_cast<PUP::able*>(t.get());
       p | t1;
     }
   }

